### PR TITLE
More tests

### PR DESCRIPTION
**Pull request by hategan**
This PR started as an effort to add more tests, particularly in the direction of resource specification/allocation.

It needed some local validation, so a test launcher was added that can fake multiple hosts.

As a side effect, ResourceSpecV1 was also revamped a bit: it made no sense that one couldn't specify, say `process_count=4` and `node_count=2`. That is now possible. ResourceSpecV1 checks basic constraints, such as `process_count % node_count = 0`, `process_count = node_count * ppn`. It can also infer some of these numbers if they are not otherwise specified. However, whether the inferred numbers are used or not by an executor is left to the executor.

### CODE REVIEW

**Comment by hategan** [comment]
Sorry. Converted the wrong PR to draft.

File: src/psij/resource_spec.py [hunk 1]
```diff
@@ -1,9 +1,17 @@
...
 from psij.exceptions import InvalidJobException
 
 
+def _nulls(objs: List[object]) -> int:
+    s = 0
+    for e in objs:
+        if e is None:
+            s += 1
+    return s
+
```
Comment (andre-merzky) [inline src/psij/resource_spec.py:13]: Isn't this the same as `list.count(value)`?
```py
>>> l = [1, None, 2, None, 3]
>>> l.count(None)
2
```

**Review by andre-merzky — APPROVED** [review]
LGTM.  Left a comment unrelated to the added tests.

File: src/psij/resource_spec.py (hunk 1 above)
Comment (hategan) [reply src/psij/resource_spec.py:13]: Silly me. Yes. Will update.

**Pull request merged**